library/core/src/array/mod.rs

```diff
@@ -503,11 +503,121 @@ impl<T, const N: usize> [T; N] {
     /// ```
     #[unstable(feature = "array_zip", issue = "80094")]
     pub fn zip<U>(self, rhs: [U; N]) -> [(T, U); N] {
-        let mut iter = IntoIterator::into_iter(self).zip(rhs);
+        self.zip_map(rhs, |a, b| (a, b))
+    }
 
-        // SAFETY: we know for certain that this iterator will yield exactly `N`
-        // items.
-        unsafe { collect_into_array_unchecked(&mut iter) }
+    /// 'Zips up' two arrays into a single array, applying the `op` over the pairs
+    ///
+    /// This is equivalent but faster than doing manual zip + map
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// #![feature(array_zip, array_zip_map)]
+    /// # use std::ops::Add;
+    /// let op = i32::add;
+    ///
+    /// let x = [1, 2, 3];
+    /// let y = [4, 5, 6];
+    ///
+    /// let output1 = x.zip(y).map(|(a, b)| op(a, b));
+    /// let output2 = x.zip_map(y, op);
+    ///
+    /// assert_eq!(output1, output2);
+    /// ```
+    #[unstable(feature = "array_zip_map", issue = "none")]
+    pub fn zip_map<U, F, R>(self, rhs: [U; N], mut op: F) -> [R; N]
+    where
+        F: FnMut(T, U) -> R,
+    {
+        // If any of the items need drop, take the 'slow' path that ensures all drop
+        // handlers are called in case of panic.
```

Comment:
Assuming this is worth doing, I think it should happen in `try_collect_into_array` instead (https://github.com/rust-lang/rust/blob/9fbbe75fd73ed3c202cec3b0ba51eadf506f03fe/library/core/src/array/mod.rs#L794), so that the other functions can benefit from doing it too

Reply:
Yeah, as I said in my other comment, it doesn't actually seem to do much here. Although my new private methods on `IntoIter` make this much simpler to implement and share the drop checking code